plugins/ipc/stipc.cpp

```diff
@@ -103,9 +151,14 @@ class headless_input_backend_t
         backend = wlr_headless_backend_create(core.display);
         wlr_multi_backend_add(core.backend, backend);
 
-        pointer  = wlr_headless_add_input_device(backend, WLR_INPUT_DEVICE_POINTER);
-        keyboard = wlr_headless_add_input_device(backend, WLR_INPUT_DEVICE_KEYBOARD);
-        touch    = wlr_headless_add_input_device(backend, WLR_INPUT_DEVICE_TOUCH);
+        pointer  = virtual_pointer_device_create(&virtual_pointer);
+        keyboard = virtual_keyboard_device_create(&virtual_keyboard);
+        touch    = virtual_touch_device_create(&virtual_touch);
+
+        wl_signal_emit_mutable(&backend->events.new_input, pointer);
```

Comment:
Why _mutable here? We should use `wlr_signal_emit_safe` here.

Reply:
This is no such function in wlroots 0.16.0. The only option is `wl_signal_emit_mutable`.